src/pydvl/influence/frameworks/torch_differentiable.py

```diff
@@ -424,3 +431,205 @@ def hessian(
+    :param krylov_dimension: The number of Krylov vectors to use for the Lanczos method.
+                             If not provided, it defaults to $min(model.num_parameters, max(2*rank_estimate + 1, 20))$.
+    :param tol: The stopping criteria for the Lanczos algorithm, which stops when the difference
+                in the approximated eigenvalue is less than `tol`. Defaults to 1e-6.
+    :param max_iter: The maximum number of iterations for the Lanczos method. If not provided, it defaults to
+                     $10*model.num_parameters$
+    :param device: The device to use for executing the hessian vector product.
+    :param eigen_computation_on_gpu: If True, tries to execute the eigen pair approximation on the provided
+                                     device via cupy implementation.
+                                     Make sure, that either your model is small enough or you use a
+                                     small rank_estimate to fit your device's memory.
+                                     If False, the eigen pair approximation is executed on the CPU by scipy wrapper to
+                                     ARPACK.
+    :param torch_dtype: if not provided, current torch default dtype is used for conversion to torch
+    :return: A `LowRankProductRepresentation` instance that contains the top (up until rank_estimate) eigenvalues
+             and corresponding eigenvectors of the Hessian.
+    """
+
+    torch_dtype = torch.get_default_dtype() if torch_dtype is None else torch_dtype
+
+    if eigen_computation_on_gpu:
+        try:
+            import cupy as cp
+            from cupyx.scipy.sparse.linalg import LinearOperator, eigsh
+            from torch.utils.dlpack import from_dlpack, to_dlpack
+        except ImportError as e:
+            raise ImportError(
+                f"Try to install missing dependencies or set eigen_computation_on_gpu to False: {e}"
+            )
+
+        if device is None:
+            raise ValueError(
+                "Without setting an explicit device, cupy is not supported"
+            )
+
+        def to_torch_conversion_function(x):
+            return from_dlpack(x.toDlpack()).to(torch_dtype)
+
+        def mv(x):
+            x = to_torch_conversion_function(x)
+            y = hessian_vp(x) + hessian_perturbation * x
+            return cp.from_dlpack(to_dlpack(y))
+
+    else:
+        from scipy.sparse.linalg import LinearOperator, eigsh
+
+        def mv(x):
+            x_torch = torch.as_tensor(x, device=device, dtype=torch_dtype)
+            y: NDArray = (
+                (hessian_vp(x_torch) + hessian_perturbation * x_torch)
+                .detach()
+                .cpu()
+                .numpy()
+            )
+            return y
+
+        to_torch_conversion_function = partial(torch.as_tensor, dtype=torch_dtype)
+
+    from scipy.sparse.linalg import ArpackNoConvergence
```

Comment:
For now, scipy is always installed so this import can be moved to the top level.

Reply:
Agree